cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i

```diff
@@ -248,11 +475,68 @@ services::Status PredictRegressionTask<algorithmFPType, cpu>::runInternal(servic
-            ReadRows<algorithmFPType, cpu> xBD(const_cast<NumericTable *>(this->_data), iStartRow, nRowsToProcess);
+            ReadRows<algorithmFPType, cpu> xBD(const_cast<NumericTable *>(_data), iStartRow, nRowsToProcess);
             DAAL_CHECK_BLOCK_STATUS_THR(xBD);
-            algorithmFPType * res = resBD.get() + iStartRow;
 
-            predict(iTree, nTreesTotal, nRowsToProcess, dim.nCols, xBD.get(), res, dim);
+            if (predShapContributions)
+            {
+                // thread-local write rows into global result buffer
+                WriteOnlyRows<algorithmFPType, cpu> resRow(result, iStartRow, nRowsToProcess);
+                DAAL_CHECK_BLOCK_STATUS_THR(resRow);
+                auto resRowPtr = resRow.get();
+                if ((predictionBias * predictionBias) > (DBL_EPSILON * DBL_EPSILON))
+                {
+                    // memory is already initialized to 0
+                    // only set it to the bias term if it's != 0
+                    for (size_t i = 0; i < nRowsToProcess; ++i)
+                    {
+                        resRowPtr[i * resultNColumns + predictionIndex] = predictionBias;
+                    }
+                }
+                // bias term: prediction - sum_i phi_i (subtraction in predictContributions)
+                predict(iTree, nTreesToUse, nRowsToProcess, xBD.get(), resRowPtr, dim, resultNColumns);
+
+                // TODO: support tree weights
+                safeStat |= predictContributions(iTree, nTreesToUse, nRowsToProcess, xBD.get(), resRowPtr, 0, 0, dim);
+            }
+            else if (predShapInteractions)
+            {
+                // thread-local write rows into global result buffer
+                WriteOnlyRows<algorithmFPType, cpu> resRow(result, iStartRow, nRowsToProcess);
+                DAAL_CHECK_BLOCK_STATUS_THR(resRow);
+
+                // nominal values are required to calculate the correct bias term
+                algorithmFPType * nominal = static_cast<algorithmFPType *>(daal_malloc(nRowsToProcess * sizeof(algorithmFPType)));
+                if (!nominal)
+                {
+                    safeStat.add(ErrorMemoryAllocationFailed);
+                    return;
+                }
+                for (size_t i = 0; i < nRowsToProcess; ++i)
+                {
+                    nominal[i] = predictionBias;
+                }
+                predict(iTree, nTreesToUse, nRowsToProcess, xBD.get(), nominal, dim, 1);
+
+                // TODO: support tree weights
+                safeStat |= predictContributionInteractions(iTree, nTreesToUse, nRowsToProcess, xBD.get(), nominal, resRow.get(), dim);
+
+                daal_free(nominal);
+            }
+            else
+            {
+                algorithmFPType * res = resMatrix.get() + iStartRow;
+                if ((predictionBias * predictionBias) > (DBL_EPSILON * DBL_EPSILON))
```

Comment:
1) Comparison operands order
2) Please use `constexpr` variable to compare with
3) (Important) `DBL_EPSILON` is extremely small. Square of this value is even smaller. There is no point in comparison with it. Please just use something like comparison with an absolute value.

Reply:
Yeah, doing DBL_EPSILON * DBL_EPSILON is completely pointless, how could I miss this.